source/extensions/injected_credentials/common/credential.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "envoy/common/exception.h"
+#include "envoy/http/header_map.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace InjectedCredentials {
+namespace Common {
+
+/**
+ * Credential injector injects credential to HTTP headers.
+ */
+class CredentialInjector {
+public:
+  virtual ~CredentialInjector() = default;
+
+  /*
+   * Request for a credential from a given source.
+   */
+  class Request {
+  public:
+    virtual ~Request() = default;
+
+    /**
+     * Signals that request should be cancelled.
+     */
+    virtual void cancel() PURE;
+  };
+  using RequestPtr = std::unique_ptr<Request>;
+
+  /**
+   * Notifies caller of the request status.
+   */
+  class Callbacks {
+  public:
+    virtual ~Callbacks() = default;
+
+    /**
+     * Called when the async request for credentials succeeds.
+     */
+    virtual void onSuccess() PURE;
```

Comment:
I think this API may cannot support the injectors that need to inject different secrets for different requests?

Like the one you noticed multiple times, oauth2 credential provider. When you send an async request to remote server, and response should only be used by the current request.

If the onSuccess have no any parameter, then it means the injector (it will be shared by multiple requests at same time) need to store all responses first, and when filter call the `inject(headers)`, the injector need to select the correct response according to the headers. I think it's hard to make it work as your expection.

Reply:
As https://github.com/envoyproxy/envoy/pull/27769 indicates, the identity is associated with the workload, not requests.

> This filter is intended to be used for workload authentication, which means that the identity associated with the inserted credential is considered as the identity of the workload behind the envoy proxy(in this case, envoy is typically deployed as a sidecar alongside that workload).

Reply:
Hah, then it should be OK to use current API. But I think in the future, there maybe more requirement to support per request credential (or other identifies).

Anyway, it's OK for now. Thanks.